sources/packages/backend/libs/integrations/src/esdc-integration/e-cert-integration/e-cert-file-handler.ts

```diff
@@ -99,97 +108,123 @@ export class ECertFileHandler extends ESDCFileHandler {
-      async (nextSequenceNumber: number, entityManager: EntityManager) => {
-        try {
-          this.logger.log(
-            `Creating  ${offeringIntensity} e-Cert file content...`,
-          );
-          const fileContent = eCertIntegrationService.createRequestContent(
-            disbursementRecords,
+    try {
+      const sequenceGroup = `${sequenceGroupPrefix}_${getISODateOnlyString(
+        new Date(),
+      )}`;
+      let uploadResult: ECertUploadResult;
+      await this.sequenceService.consumeNextSequence(
+        sequenceGroup,
+        async (nextSequenceNumber: number, entityManager: EntityManager) => {
+          uploadResult = await this.processECert(
             nextSequenceNumber,
-          );
-
-          // Create the request filename with the file path for the e-Cert File.
-          const fileInfo = await this.createRequestFileName(
+            entityManager,
+            eCertIntegrationService,
+            offeringIntensity,
             fileCode,
-            nextSequenceNumber,
           );
+        },
+      );
+      return uploadResult;
+    } catch (error: unknown) {
+      if (
+        error instanceof CustomNamedError &&
+        error.name === ECERT_GENERATION_NO_RECORDS_AVAILABLE
+      ) {
+        return {
+          generatedFile: "none",
+          uploadedRecords: 0,
+        };
+      }
+      throw error;
+    }
+  }
 
-          // Creates the repository based on the entity manager that
-          // holds the transaction already created to manage the
-          // sequence number.
-          const disbursementScheduleRepo =
-            entityManager.getRepository(DisbursementSchedule);
-          await this.disbursementScheduleService.updateRecordsInSentFile(
-            disbursementIds,
-            now,
-            disbursementScheduleRepo,
-          );
+  /**
+   * Prepare the disbursements for e-Cert generation and upload the e-Cert file.
+   * @param sequenceNumber e-Cert sequence number.
+   * @param entityManager manages the current DB transaction.
+   * @param eCertIntegrationService Full-Time/Part-Time integration responsible
+   * for the respective integration.
+   * @param offeringIntensity disbursement offering intensity.
+   * @param fileCode file code applicable for Part-Time or Full-Time.
+   * @returns information of the uploaded e-Cert file.
+   * @throws CustomNamedError ECERT_GENERATION_NO_RECORDS_AVAILABLE
+   */
+  private async processECert(
+    sequenceNumber: number,
+    entityManager: EntityManager,
+    eCertIntegrationService: ECertIntegrationService,
+    offeringIntensity: OfferingIntensity,
+    fileCode: string,
+  ): Promise<ECertUploadResult> {
+    try {
+      const disbursements =
+        await this.eCertGenerationService.prepareDisbursementsForECertGeneration(
+          offeringIntensity,
+          entityManager,
+        );
+      if (!disbursements.length) {
+        throw new CustomNamedError(
```

Comment:
Why do we need to throw error when no records are available?

Reply:
All this processing is happening inside the  `consumeNextSequence` and it also executes the lock on DB to avoid two e-Cert generations happening at the same time. Part-time/full-time can happen in parallel and the lock is different because each process uses a specific sequence group prefix (ECERT_FT_SENT_FILE/ECERT_PT_SENT_FILE).
The `consumeNextSequence` is also providing the `entityManager` that is used to execute all DB updates in a single transaction.
The method `prepareDisbursementsForECertGeneration` is also entirely using the transaction, so even to execute the first selection to check for e-Cert records to be generated the DB lock is first acquired. Since the transaction is managed by the `consumeNextSequence`, throwing an exception at this level is a way to let the `consumeNextSequence` be aware that it must rollback the transaction and release the lock. Please let me know if we need to talk further about it.

![image](https://user-images.githubusercontent.com/61259237/211691139-18b7f6ff-d13d-4d29-bdbb-f206c2205595.png)